src/transport/SecureSessionTable.h

```diff
@@ -133,21 +91,99 @@ class SecureSessionTable
      * @return the session if found, NullOptional if not found
      */
     CHECK_RETURN_VALUE
-    Optional<SessionHandle> FindSecureSessionByLocalKey(uint16_t localSessionId)
-    {
-        SecureSession * result = nullptr;
-        mEntries.ForEachActiveObject([&](auto session) {
-            if (session->GetLocalSessionId() == localSessionId)
-            {
-                result = session;
-                return Loop::Break;
-            }
-            return Loop::Continue;
-        });
-        return result != nullptr ? MakeOptional<SessionHandle>(*result) : Optional<SessionHandle>::Missing();
-    }
+    Optional<SessionHandle> FindSecureSessionByLocalKey(uint16_t localSessionId);
 
 private:
+    friend class TestSecureSessionTable;
+
+    /**
+     * This provides a sortable wrapper for a SecureSession object. A SecureSession
+     * isn't directly sortable since it is not swappable (i.e meet criteria for ValueSwappable).
+     *
+     * However, this wrapper has a stable pointer to a SecureSession while being swappable with
+     * another instance of it.
+     *
+     */
+    struct SortableSession
+    {
+    public:
+        void swap(SortableSession & other)
+        {
+            SortableSession tmp(other);
+            other.mSession = mSession;
+            mSession       = tmp.mSession;
+        }
+
+        const Transport::SecureSession * operator->() const { return mSession; }
+
+    private:
+        SecureSession * mSession;
+        friend class SecureSessionTable;
+    };
+
+    /**
+     *
+     * Encapsulates all the necessary context for an eviction policy callback
+     * to implement its specific policy. The context is provided to the callee
+     * with the expectation that it'll call Sort() with a comparator function provided
+     * to get the list of sessions sorted in the desired order.
+     *
+     */
+    class EvictionPolicyContext
+    {
+    public:
+        /*
+         * Called by the policy implementor to sort the list of sessions given a comparator
+         * function. The provided function shall have the following signature:
+         *
+         * bool CompareFunc(const SortableSession &a, const SortableSession &b);
+         *
+         * If a is a better candidate than b, true should be returned. Else, return false.
+         *
+         * NOTE: Sort() can be called multiple times.
+         *
+         */
+        template <typename CompareFunc>
+        void Sort(CompareFunc func)
+        {
+            std::sort(mSessionList.begin(), mSessionList.begin() + mSessionList.size(), func);
```

Comment:
How much codesize is this ending up pulling in?  And is our session pool going to be big enough to make that worth it, compared to a simple bubblesort or whatnot?

Reply:
NRF is showing around a 1.2K flash increase for .text, and other embedded platforms are in the similar range. 1.2K doesn't seem crazy excessive...

That being said, I don't think our session pool is going to be that big that it justifies the performance improvements there. Will either try to use `qsort` (which I believe has better flash characteristics that std::sort), or a custom impl of bubble sort.

Reply:
qsort will likely be smaller, as std::sort probably is a quick sort + some other algorithm for large sets.

Custom rolling sort code does not seem like a good choice to save bytes.

Reply:
I ended up going down the path of hand-rolling a bubble sort since it's super easy to implement. If you feel strongly about that, in the next PR, I can switch over to `qsort` to see the delta in flash and see if that is acceptable.